pkg/serverless/trace/propagation/carriers.go

```diff
@@ -126,32 +139,62 @@ func sqsMessageAttrCarrier(attr events.SQSMessageAttribute) (tracer.TextMapReade
 	return carrier, nil
 }
 
+// snsBody is used to  unmarshal only required fields on events.SNSEntity
+// types.
+type snsBody struct {
+	MessageAttributes map[string]interface{}
+}
+
 // snsSqsMessageCarrier returns the tracer.TextMapReader used to extract trace
 // context from the body of an events.SQSMessage type.
 func snsSqsMessageCarrier(event events.SQSMessage) (tracer.TextMapReader, error) {
-	var body struct {
-		MessageAttributes map[string]struct {
-			Type  string
-			Value string
-		}
-	}
+	var body snsBody
 	err := json.Unmarshal([]byte(event.Body), &body)
 	if err != nil {
 		return nil, fmt.Errorf("Error unmarshaling message body: %w", err)
 	}
-	msgAttrs, ok := body.MessageAttributes[datadogSQSHeader]
+	return snsEntityCarrier(events.SNSEntity{
+		MessageAttributes: body.MessageAttributes,
```

Comment:
Just want to confirm if this mapping relationship is the same for raw message delivery?

Reply:
nvm, I think I understand it now.This is quite elegant.